src/olympia/accounts/tasks.py

```diff
@@ -10,24 +11,45 @@
 log = olympia.core.logger.getLogger('z.accounts')
 
 
+def user_profile_from_uid(f):
+    @functools.wraps(f)
+    def wrapper(uid, timestamp, *args, **kw):
+        try:
+            timestamp = datetime.fromtimestamp(timestamp)
+            profile = UserProfile.objects.get(fxa_id=uid)
+            return f(profile, timestamp, *args, **kw)
+        except ValueError as e:
+            log.warning(e)
+        except UserProfile.MultipleObjectsReturned:
+            log.warning('Multiple profile matches for FxA id %s' % uid)
+        except UserProfile.DoesNotExist:
+            log.info('No profile match for FxA id %s' % uid)
+    return wrapper
+
+
 @task
 @use_primary_db
-def primary_email_change_event(email, uid, timestamp):
+@user_profile_from_uid
+def primary_email_change_event(profile, changed_date, email):
     """Process the primaryEmailChangedEvent."""
-    try:
-        profile = UserProfile.objects.get(fxa_id=uid)
-        changed_date = datetime.fromtimestamp(timestamp)
-        if (not profile.email_changed or
-                profile.email_changed < changed_date):
-            profile.update(email=email, email_changed=changed_date)
-            log.info('Account [%s] email [%s] changed from FxA on %s' %
-                     (profile.id, email, changed_date))
-        else:
-            log.warning('Account [%s] email updated ignored, %s > %s' %
-                        (profile.id, profile.email_changed, changed_date))
-    except ValueError as e:
-        log.warning(e)
-    except UserProfile.MultipleObjectsReturned:
-        log.warning('Multiple profile matches for FxA id %s' % uid)
-    except UserProfile.DoesNotExist:
-        log.info('No profile match for FxA id %s' % uid)
+    if (not profile.email_changed or
+            profile.email_changed < changed_date):
+        profile.update(email=email, email_changed=changed_date)
+        log.info(
+            'Account pk [%s] email [%s] changed from FxA on %s' % (
+                profile.id, email, changed_date))
+    else:
+        log.warning('Account pk [%s] email updated ignored, %s > %s' %
+                    (profile.id, profile.email_changed, changed_date))
+
+
+@task
+@use_primary_db
+@user_profile_from_uid
+def delete_user_event(user, deleted_date):
```

Comment:
Trying to understand how the `@user_profile_from_uid` decorator works. In this case `user` is what is returned by the wrapper in the decorator, which is the same thing as `profile` in `primary_email_change_event`, above (just a different argument name)? So it's positional - i.e., the first argument to this function is whatever is returned by `@user_profile_from_uid`?

Reply:
`return f(profile, timestamp, *args, **kw)` is the return from the decorator, so yeah, it takes a user arg (which is then an fxa_id) and a date-time arg (which is then a timestamp) and returns the same but now as a `UserProfile` and a `datetime` instance.

user or profile... it's a `UserProfile` instance so either variable name is descriptive.